games/xworld/maps/xworld_env.py

```diff
@@ -409,30 +428,36 @@ def __instantiate_entities(self):
         after which its properties are set.
         The entities should have been set in _configure()
         """
-        Y, X = self.get_dims()
-
-        maze = spanning_tree_maze_generator(X, Y)
-        blocks = [(j, i) for i,m in enumerate(maze) for j,b in enumerate(m) if b == '#']
-
-        ## maybe not all blocks of the maze will be used later
-        random.shuffle(blocks)
-
-        ## first remove all maze blocks from the available set
-        for b in blocks:
-            if b in self.available_grids:
-                self.available_grids.remove(b)
-
-        ## select a random img path for each entity
-        for i, e in enumerate(self.entities):
-            self.set_property(e)
-
-        ## add back some empty grids
-        self.available_grids += blocks[len(self.get_blocks()):]
-
-        ## use the remaining blocks
-        for e in self.get_blocks():
-            assert blocks, "too many blocks for a valid maze"
-            e.loc = blocks.pop()
+        if self.maze_generation:
+            Y, X = self.get_dims()
+            maze = spanning_tree_maze_generator(X, Y)
+            blocks = [(j, i) for i,m in enumerate(maze) for j,b in enumerate(m) if b == '#']
+
+            ## maybe not all blocks of the maze will be used later
+            random.shuffle(blocks)
+
+            ## first remove all maze blocks from the available set
+            for b in blocks:
+                if b in self.available_grids:
+                    self.available_grids.remove(b)
+            ## instantiate properties for each entity
+            for i, e in enumerate(self.entities):
+                if e.loc is not None:
+                    warnings.warn("Maze generation is on! Over-writing pre-specified location %s!" % (e.loc,))
+                    e.loc = None # remove the pre-set location when maze_generation is on
+                self.set_property(e)
+            ## add back some empty grids
+            self.available_grids += blocks[len(self.get_blocks()):]
+        else:
+            ## instantiate properties for each entity
+            for i, e in enumerate(self.entities):
+                self.set_property(e)
+
+        if self.maze_generation:
```

Comment:
merge this code block into the previous branch

Reply:
merged